File: performance-comparison/benchmarks/haskell-project/src/HashMap.hs
-- hash_map -- insert i -> i*2, look every key back up, sum.  Data.IntMap
-- (containers, a GHC boot package) is the analogous persistent map to the
-- Turmeric HAMT column -- noted in docs/methodology.md.
{-# LANGUAGE BangPatterns #-}
import System.Environment (getArgs)
import qualified Data.IntMap.Strict as IM
import Data.List (foldl')
main :: IO ()
main = do
  args <- getArgs
  let n = case args of (a:_) -> read a; _ -> 1000 :: Int
      m = foldl' (\acc i -> IM.insert i (i * 2) acc) IM.empty [0 .. n - 1]
      s = foldl' (\acc i -> acc + IM.findWithDefault 0 i m) 0 [0 .. n - 1]
  print s

File: performance-comparison/benchmarks/rust-workspace/src/bin/hash_map.rs
// hash_map -- insert i -> i*2 for i in [0, n), then sum every value back
// out by key lookup. std::collections::HashMap is Rust's analogous map to
// the Turmeric HAMT / Haskell IntMap columns (docs/methodology.md).
use std::collections::HashMap;
use std::env;

fn hashmap_bench(n: i64) -> i64 {
    let mut m: HashMap<i64, i64> = HashMap::new();
    for i in 0..n {
        m.insert(i, i * 2);
    }
    let mut sum: i64 = 0;
    for i in 0..n {
        sum += *m.get(&i).unwrap_or(&0);
    }
    sum
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(1000);
    println!("{}", hashmap_bench(n));
}
